# How Floats work (in a bit more depth)

Let's start the discussion by stating that computers- at the fundamental level- work by flicking on and off trillions of little switches (transistors). When Python receives a number (in any base) it converts that number into base 2 (binary). By converting the number to a bunch of one's and zero's it makes it easier for the machine to calculate.

So, at the fundamental level, Python is working out sums in base 2 and we puny mortals are thinking in base 10.  Long story short, Python will *round* numbers that cannot be easily expressed in base 2 but interestingly enough of those sums are easy in base 10. The net effect is that there are going to be cases where we feel python makes an “obvious” error in arithmetic; it might seem like a bug when it happens. But actually, these ‘quirks’ are just a consequence of how floats are implemented at the low level.  

Let's take a look at a few examples...

In [6]:
print (0.1 + 0.1)

0.2


0.1 + 0.1 = 0.2 Okay, no surprises here. But what about the following sum:

In [7]:
print(0.1 + 0.1 + 0.1)

0.30000000000000004


This sum should equal 0.3, you don’t need to be a genius to figure that out. And we can see from what Python return its close to right answer but  nonetheless wrong. What happened? well 0.1 * 3 is easy for us to do in base 10 but in base 2 its a bit more tricky. Instead of trying to find the actual value Python "gives up" after a while and just does a bit of rounding.

Similarly...

In [8]:
print(0.7 + 0.7 + 0.7)

2.0999999999999996


2.099... isn't what we would expect (2.1)

Alright, that just about wraps up this lecture on numbers. If you would like an even more in-depth explanation how floats works please [read this](https://docs.python.org/3/tutorial/floatingpoint.html).